[DataGrid] Add fluid columns width support (#480)
mui-org/material-ui-x: docs/src/pages/components/data-grid/columns/columns.md

{{"demo": "pages/components/data-grid/columns/ColumnWidthGrid.js", "bg": "inline"}}

### Fluid width

> By default each column have a fixed width of 100 pixels, but column fluidity (responsiveness) can be by achieved setting the `flex` propery avaibale in the `ColDef`.

Reviewer: ```suggestion
Each column has a fixed width of 100 pixels by default, but column fluidity (responsiveness) can be by achieved by setting the `flex` property in `ColDef`.
```

> The `flex` propery accepts values between 0 and ∞.

Reviewer: ```suggestion
The `flex` property accepts a value between 0 and ∞.
```

> The `flex` property works by dividing the remaining space in the grid among all flex columns in proportion to their `flex` value. For example, a grid with a total width of 500px that has three columns: the first with width: 200; the second with flex: 1; and third with flex: 0.5. The first column will be 200px wide, leaving 300px remaining. The column with flex: 1 has twice the size with flex: 0.5 which means that final sizes will be: 200px, 200px, 100px.

Reviewer: ```suggestion
The `flex` property works by dividing the remaining space in the grid among all flex columns in proportion to their `flex` value. 
For example, consider a grid with a total width of 500px that has three columns: the first with `width: 200`; the second with `flex: 1`; and third with `flex: 0.5`. 
The first column will be 200px wide, leaving 300px remaining. The column with `flex: 1` is twice the size of `flex: 0.5`, which means that final sizes will be: 200px, 200px, 100px.
```

> The `flex` configuration does not work together with the `width` configuration. If you set both the `flex` and `width` inside the `ColDef` of column the `flex` will overwrite the `width`.

Reviewer: ```suggestion
Note that `flex` doesn't work together with `width`. If you set both `flex` and `width` in `ColDef`, `flex` will override `width`.
```

> The `flex` configuration does not work if the combined width of the columns that have `width` is more that the width of the grid itself. If that is the case a scroll bar will be visible and the columns that hava `flex` will default back to their base value of 100 pixels.

Reviewer 2: typo:  ...is more than* the width of the grid itself.

Author: Thanks

> The `flex` configuration does not work if the combined width of the columns that have `width` is more than the width of the grid itself. If that is the case a scroll bar will be visible and the columns that hava `flex` will default back to their base value of 100 pixels.

Reviewer: ```suggestion
In addition, `flex` does not work if the combined width of the columns that have `width` is more than the width of the grid itself. If that is the case a scroll bar will be visible and the columns that have `flex` will default back to their base value of 100px.
```

{{"demo": "pages/components/data-grid/columns/ColumnWidthGrid.js", "bg": "inline"}}

> ## Column fluid width

Reviewer 3: I think that it should be part of the generic column width concern. 

```suggestion
### Fluid width
```

Reviewer 3: I wonder about 
```suggestion
### Flex width
```

Author: I was wondering how to structure it but, yes having it as a subsection of the `Column width` does look better.

Reviewer 3: I'm raising it because I want on using a flatter structure than ag-grid for the documentation. We will probably in the future need to break it further down as we add more features to the component.

> By default each column have a fixed width of 100 pixels but we can have fluid (responsive) columns by setting the `flex` propery avaibale in the `ColDef`.

Reviewer 3: The usage of "we" is forbidden in the documentation (or at least limited to rare exceptions) https://trello.com/c/1Q7hqXHu/2714-technical-documentation-guidelines

Author: My bad, sorry about that, I'll fix it now.

Reviewer 3: No problem! I'm doing this too from time to time. Matt is rectifying me when I do it wrong.

The `flex` propery accepts values between 0 and 1.

The `flex` property works by dividing the remaining space in the grid among all flex columns in proportion to their `flex` value. For example, a grid with a total width of 500px that has three columns: the first with width: 200; the second with flex: 1; and third with flex: 0.5. The first column will be 200px wide, leaving 300px remaining. The column with flex: 1 has twice the size with flex: 0.5 which means that final sizes will be: 200px, 200px, 100px.

The `flex` configuration does not work together with the `width` configuration. If you set both the `flex` and `width` inside the `ColDef` of column the `flex` will overwrite the `width`.

The `flex` configuration does not work if the combined width of the columns that have `width` is more that the width of the grid itself. If that is the case a scroll bar will be visible and the columns that hava `flex` will default back to their base value of 100 pixels.

{{"demo": "pages/components/data-grid/columns/ColumnFluidWidthGrid.js", "bg": "inline"}}

> <!--

Reviewer 3: I think that we can remove these comments. We can include the benchmark in the related issue.

Author: These links are already in the ticket itself. I saw that we also added some benchmark links in the `resizing` section so that why I added these once but yes without them it looks cleaner ;)

Reviewer 3: Definitely, I was hesitating about where it would be best to store these benchmarks. What you see in the resizing section is something I forgot to clean up :). I came to the conclusion that the list of a benchmark is living content. it would be too slow to edit them in GIT. It might also create issues if we translate the page.

Note that `flex` doesn't work together with `width`. If you set both `flex` and `width` in `ColDef`, `flex` will override `width`.

> In addition, `flex` does not work if the combined width of the columns that have `width` is more than the width of the grid itself. If that is the case a scroll bar will be visible and the columns that have `flex` will default back to their base value of 100px.

Reviewer: ```suggestion
In addition, `flex` does not work if the combined width of the columns that have `width` is more than the width of the grid itself. If that is the case a scroll bar will be visible, and the columns that have `flex` will default back to their base value of 100px.
```
